### DISCUSSION TITLE
NVIDIA Flow Licesing Question

### DISCUSSION THREAD

**Discussion by DRKV333** [NVIDIA-Omniverse/PhysX#168] [Q&A]
I am a little bit confused about the licensing of the included Flow library. As far as I can tell, most of the source code for Flow is not included in this repository, only headers, some helpers, and a demo application. Crucially all the HLSL shaders are missing. These are only provided in binary form in the `_build` folder.

My main question is regarding these binaries. Are these still provided under the 3-Clause BSD license? https://github.com/NVIDIA-Omniverse/PhysX/blob/8f6cf9cfe89f2bedafbf7788a72cb04f11c31e1f/flow/PACKAGE-LICENSES/nvflow-LICENSE.md

My current guess is that they are, since that is the only license I could find in the repo. I find this a little weird though, since the license is supposed to let me redistribute Flow `with or without modification`, but when I look at the shaders inside the binary, to potentially modify or redistribute them, I see this:
```
/*
 * Copyright (c) 2014-2020, NVIDIA CORPORATION.  All rights reserved.
 *
 * NVIDIA CORPORATION and its licensors retain all intellectual property
 * and proprietary rights in and to this software, related documentation
 * and any modifications thereto.  Any use, reproduction, disclosure or
 * distribution of this software and related documentation without an express
 * license agreement from NVIDIA CORPORATION is strictly prohibited.
 */
```
Which seems to contradict the license. Or does the 3-Clause BSD license count as `an express license agreement from NVIDIA CORPORATION`? Surely this can't be right, cause then anyone could just reverse engineer the binaries, and redistribute a source code version of the library.

Are there any plans in the future, to provide the source code for the entire Flow library?

**Comment by preist-nvidia** [comment]
Converting to Q&A discussion. Will check and get back to you asap.

**Reply by preist-nvidia** [reply]
Hi @DRKV333 - I cannot find any reference to the proprietary license text in the generated hlsl files. E.g. `flow\_generated\windows-x86_64\release\nvfloweditor\shaders\ImguiBuildCS_separateSampler.hlsl` has (twice) the BSD license. Where did you see it?

Note that we accidentally had the wrong license in some places in an early version, which was fixed in https://github.com/NVIDIA-Omniverse/PhysX/commit/b2aad1b2bc0b145d3c97b246b783c9694a27da0d

Philipp

**Reply by DRKV333** [reply]
Hi Philipp, thanks for the quick reply.

The proprietary license is not in any of the provided source files, or anything derived from them. It's in the binary `nvflow.dll`. It appears several times, for instance, in `_build\windows-x86_64\debug\nvflow.dll` at file offset `0x133E54`. I'm not exactly sure about which chunk of shader text in the dll corresponds to what source file, since the source files are not in this repo.

If you've also updated the license header in the unreleased source files when b2aad1b2bc0b145d3c97b246b783c9694a27da0d happened, then I think the issue is that the dlls in `_build` have not been updated since the initial commit (80c0aab63e67fcc269c849bbfc2ac76bee4a129a).

**Reply by preist-nvidia** [reply]
I understand, and yes I think we did not rebuild the dll hence the old license there. I'll check with the team about what to do here.

**Comment by DRKV333** [comment]
I've also discovered another, unrelated problem with the Flow binaries. It appears that `windows-x86_64\debug\nvflow.pdb` and `windows-x86_64\debug\nvflowext.pdb` where compiled with `/DEBUG:FASTLINK`, which means they can't be used without all their corresponding `.obj` files. The rest of the `.pdb` files seem fine.

**Comment by DRKV333** [comment]
Hey @preist-nvidia, do you have an update regarding any of my initial questions? What is the intended licensing situation of the Flow binaries? I feel like this would be important to clear up, since not knowing what the license actually is makes it rather impossible to use the library for anyone. Plus forks might be unintentionally redistributing proprietary NVIDIA code.

**Reply by preist-nvidia** [reply]
Hi @DRKV333 - I don't have an update yet, but will post asap.
